### PR TITLE
Multiple Outline Groups

### PR DESCRIPTION
**Pull request by Zvork**
This PR adds multiple outline groups with each group having a different outline style (color, width, etc...). Each outline group is linked to a different separate selection. In this first proposal, the number of outline groups is set to 8 but can easily be increased by changing an enumeration value in the render scene before compilation.

# Test plan
Please refer to the _tests/engine/render/effects/outline/coverage_ test plan in the [hifi_tests repository ](https://github.com/highfidelity/hifi_tests/blob/master/tests/engine/render/effect/outline/coverage/test.md) for the manual or automated test plan.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4342/HighFidelity-Beta-PR11650-43f9db5b70ea2aa025948ee1e9e693984f09d2e4.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4342/HighFidelity-Beta-PR11650-43f9db5b70ea2aa025948ee1e9e693984f09d2e4.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4343/HighFidelity-Beta-PR11650-7011fd3731ef9bf10217deac219604bf1d2314f6.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4343/HighFidelity-Beta-PR11650-7011fd3731ef9bf10217deac219604bf1d2314f6.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4344/HighFidelity-Beta-PR11650-68ed061ce59ae59a5c88444b7f3349ae02967b40.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4344/HighFidelity-Beta-PR11650-68ed061ce59ae59a5c88444b7f3349ae02967b40.exe

**Comment by YuppyPlays** [comment]
Testing.
[Finally found the test plan!]

**Comment by YuppyPlays** [comment]
Testing complete:

Can confirm from Test plan located: https://github.com/Zvork/hifi_tests/blob/63212db8d933d7f42b579c9f37cf32c92f113636/tests/engine/render/effects/outline/test.md
All 13 steps come out as expected by the test plans pictures, used the "Space key" as the script automates through each step for visual comparison, all **Pass**.

**Comment by RebeccaStankus** [comment]
I looks like outlines are showing incorrectly in HMD mode with this PR. Everything works as expected in desktop mode but outlines in HMD appear in the wrong places, often in zig zag lines, and sometimes are only visible depending on the angle at which they are viewed. If you run the test plan in HMD, you can see the issues.

**Comment by Zvork** [comment]
@RebeccaStankus I've disabled the optimization that broke the outline in HMD mode and I've also updated the test case in the hifi_tests PR to work in an empty domain with generated entities. It is now fully independant and automatic.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4503/HighFidelity-Beta-PR11650-4cadcd79bfb522e453a75fa46f859ccbcdcd6d27.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4503/HighFidelity-Beta-PR11650-4cadcd79bfb522e453a75fa46f859ccbcdcd6d27.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4505/HighFidelity-Beta-PR11650-4cbee72eae60305cc616b53aee5fba092f2090ae.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4505/HighFidelity-Beta-PR11650-4cbee72eae60305cc616b53aee5fba092f2090ae.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4523/HighFidelity-Beta-PR11650-3a977f60397077aa2a7f18712c41736c6b354f24.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4523/HighFidelity-Beta-PR11650-3a977f60397077aa2a7f18712c41736c6b354f24.exe

**Comment by ZappoMan** [comment]
needs to be retested... code changed since last QA Pass....

File: interface/src/scripting/SelectionScriptingInterface.h [hunk 1]
```diff
@@ -61,6 +61,7 @@ class SelectionScriptingInterface : public QObject, public Dependency {
 
     Q_INVOKABLE bool addToSelectedItemsList(const QString& listName, const QString& itemType, const QUuid& id);
     Q_INVOKABLE bool removeFromSelectedItemsList(const QString& listName, const QString& itemType, const QUuid& id);
+    Q_INVOKABLE bool clearSelectedItemsList(const QString& listName);
```
Comment (ZappoMan) [inline interface/src/scripting/SelectionScriptingInterface.h]: let's add jsdoc comments to this scripting interface for all the APIs available.

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Can somebody confirm if test plan is still:
https://github.com/Zvork/hifi_tests/blob/63212db8d933d7f42b579c9f37cf32c92f113636/tests/engine/render/effects/outline/test.md

I cannot seem to find anything else.

**Comment by Zvork** [comment]
Sorry I forgot to update the link following some changes request by Sam. The correct test folder is tests/engine/render/effects/outline/coverage in the "outline" branch of my Github Zvork/hifi_tests

**Comment by YuppyPlays** [comment]
Testing.

Test plan being used:
https://github.com/Zvork/hifi_tests/blob/outline/tests/engine/render/effect/outline/coverage/test.md

**Comment by YuppyPlays** [comment]
Testing complete:

**Failed** Instructions are not clear enough given what I'm experiencing, it says to use an empty area in domain but everytime I use this script in an empty far-off location, it teleports back to just above the domains starting point to run the script, giving results to shading like this:

image

**Comment by Zvork** [comment]
Right. My fault. The instructions should be "in an empty domain" not just an empty region. I will correct that once I get back.

**Comment by YuppyPlays** [comment]
@zvork No worries, I will test and see if I can make my domain empty (I've not done this before, but will see if I can figure it out quick)

Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

**[I was able to get my domain to be completely empty by deleting models.json.gz from one of the appdata folders for high fidelity.]**

Unfortunately after entering Step 3 and quickly getting this picture which shows a difference to test plan expected results (In test plan red sphere also gets the outer-orange line, where as in PR I am seeing no orange around the red sphere), I cannot seem to go any further, interface crashes completely and even retrying keeps doing so between first step and step 3.


step3

**Fail**

[Logs]:
[Logs.zip](https://github.com/highfidelity/hifi/files/1429316/Logs.zip)

**Comment by Nex-Pro** [comment]
### PASS 
Ran this on win 10 i7 with gtx1070

At each step the results were as expected. Correct line width, selection and effect each time. No issues seen.

**Step 3 image**
y34708-hifi-snap-by--on-2017-10-31_03-28-43

**Comment by Nex-Pro** [comment]
Content loading isn't relevant to this PR. Focus is on the outline rendering which all worked as expected.

**Comment by Nex-Pro** [comment]
Was due to ... [DEBUG] [hifi.networking] Server unavailable for QUrl(http://headache.hungry.com/~seth/hifi/grass.png)

**Comment by Nex-Pro** [comment]
**Step 1**
r36288-hifi-snap-by--on-2017-10-31_03-47-37

**Step 2**
c36288-hifi-snap-by--on-2017-10-31_03-47-54

**Step 3**
e36288-hifi-snap-by--on-2017-10-31_03-48-00

**Step 4**
k36288-hifi-snap-by--on-2017-10-31_03-48-07

**Step 5**
r36288-hifi-snap-by--on-2017-10-31_03-48-12

**Step 6**
u36288-hifi-snap-by--on-2017-10-31_03-48-18

**Step 7**
f36288-hifi-snap-by--on-2017-10-31_03-48-23

**Step 8**
v36288-hifi-snap-by--on-2017-10-31_03-48-27

**Step 9**
s36288-hifi-snap-by--on-2017-10-31_03-48-30

**Step 10**
d36288-hifi-snap-by--on-2017-10-31_03-48-34

**Step 11**
b36288-hifi-snap-by--on-2017-10-31_03-48-38

**Step 12**
s36288-hifi-snap-by--on-2017-10-31_03-48-42

**Comment by misslivirose** [comment]
Is it expected that this would work for Overlays?

**Comment by Zvork** [comment]
@misslivirose It isn't expected to work with Overlays but it would be very easy to do so.

**Comment by Zvork** [comment]
@YuppyPlays Would it be possible to have some information on your testing platform? Operating system, GPU brand and version, etc?

**Comment by Zvork** [comment]
@YuppyPlays Would it also be possible for you to relaunch the test after typing in the Interface console:
`Render.getConfig("RenderMainView.OutlineDebug").viewMask = true`
And show the result for step 3?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4632/HighFidelity-Beta-PR11650-e9743d4c950a1e4591ec2690b12d9aa1de70561a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4632/HighFidelity-Beta-PR11650-e9743d4c950a1e4591ec2690b12d9aa1de70561a.exe

**Comment by Zvork** [comment]
Also, don't forget to run the test with an HMD.

**Comment by Zvork** [comment]
Please retest this as some optimizations have been added. Test plan has also been modified. And of course, please test this on HMD if possible.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4672/HighFidelity-Beta-PR11650-e0fd71f24636d884d704f4dc568f98643e36dca9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4672/HighFidelity-Beta-PR11650-e0fd71f24636d884d704f4dc568f98643e36dca9.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4724/HighFidelity-Beta-PR11650-fe79730012da39e024b86af80cb60dc5fc462593.dmg

**Comment by Nex-Pro** [comment]
Awaiting update

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4724/HighFidelity-Beta-PR11650-fe79730012da39e024b86af80cb60dc5fc462593.exe

**Comment by Nex-Pro** [comment]
Testing win10

**Comment by Nex-Pro** [comment]
Using build fe79

**Comment by RebeccaStankus** [comment]
The outlines are not showing correctly in HMD again after the latest build with optimizations added.

**Comment by Nex-Pro** [comment]
### FAIL

This works fine in desktop mode apart from last stage where a large amount of flicker was apparent.  (see image sequence ) 

but I concur with @RebeccaStankus that this is totally broken in HMD. 

I had issues with the Script dialog but was able to determine the settings by the small amount of text protruding beneath each slider ( I did not feel that this dialog issue was relevant to the outcome of this test ) 

### DESKTOP image sequence.
j26176-hifi-snap-by-rocketman-on-2017-11-07_19-09-57


u26176-hifi-snap-by-rocketman-on-2017-11-07_19-10-12



b21328-hifi-snap-by-rocketman-on-2017-11-07_18-51-45



s21328-hifi-snap-by-rocketman-on-2017-11-07_18-54-14



i21328-hifi-snap-by-rocketman-on-2017-11-07_18-55-05



u21328-hifi-snap-by-rocketman-on-2017-11-07_18-55-53



a21328-hifi-snap-by-rocketman-on-2017-11-07_18-57-10



e21328-hifi-snap-by-rocketman-on-2017-11-07_18-57-16



c21328-hifi-snap-by-rocketman-on-2017-11-07_18-58-03



m21328-hifi-snap-by-rocketman-on-2017-11-07_18-59-52



d21328-hifi-snap-by-rocketman-on-2017-11-07_19-01-16



g21328-hifi-snap-by-rocketman-on-2017-11-07_19-04-10



r21328-hifi-snap-by-rocketman-on-2017-11-07_19-05-53
m21328-hifi-snap-by-rocketman-on-2017-11-07_19-05-46
z21328-hifi-snap-by-rocketman-on-2017-11-07_19-05-37
j21328-hifi-snap-by-rocketman-on-2017-11-07_19-05-10

**Comment by Nex-Pro** [comment]
@Zvork Good progress thanks.

**Comment by Zvork** [comment]
I've identified and fixed a graphical bug (the bizarre blocky rendering in your last screenshot @Nex-Pro by the look of it) and will push the fix along with the HMD corrections.

**Comment by Nex-Pro** [comment]
@Zvork Thanks

**Comment by Zvork** [comment]
@Nex-Pro , @RebeccaStankus Outline should now work with HMD. Weird rendering artifacts in corners should also be fixed. 

**PLEASE NOTE** that "outline" has now been renamed to "highlight", as recommended by @samcake . Thus the debug script for outlining / highlighting is now _scripts/developer/utilities/render/debugHighlight.js_

# TEST PLAN
Please use the updated semi-automated test case as described [here ](https://github.com/Zvork/hifi_tests/tree/highlight/tests/engine/render/effect/highlight/coverage/test.md). The test case files can be found in my repository [https://github.com/Zvork/hifi_tests/tree/highlight/tests/engine/render/effect/highlight/coverage](https://github.com/Zvork/hifi_tests/tree/highlight/tests/engine/render/effect/highlight/coverage)

**Comment by marko8904** [comment]
build this please

**Comment by Nex-Pro** [comment]
Awaiting Win build

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4761/HighFidelity-Beta-PR11650-350993bef0979b9a50e5f3356d15d46627417d01.exe

**Comment by Nex-Pro** [comment]
Testing

**Comment by Nex-Pro** [comment]
### PASS
tested on win10 and Vive

hifi-snap-by--on-2017-11-08_20-15-52
hifi-snap-by--on-2017-11-08_20-16-06
hifi-snap-by--on-2017-11-08_20-16-18
hifi-snap-by--on-2017-11-08_20-16-36
hifi-snap-by--on-2017-11-08_20-16-54
hifi-snap-by--on-2017-11-08_20-17-07
hifi-snap-by--on-2017-11-08_20-17-20
hifi-snap-by--on-2017-11-08_20-17-32
hifi-snap-by--on-2017-11-08_20-17-45
hifi-snap-by--on-2017-11-08_20-17-58
hifi-snap-by--on-2017-11-08_20-18-14
hifi-snap-by--on-2017-11-08_20-18-25
hifi-snap-by--on-2017-11-08_20-18-40
hifi-snap-by--on-2017-11-08_20-19-02
hifi-snap-by--on-2017-11-08_20-20-20
hifi-snap-by--on-2017-11-08_20-20-48
hifi-snap-by--on-2017-11-08_20-21-03


294020-hifi-snap-by--on-2017-11-08_21-04-24
294020-hifi-snap-by--on-2017-11-08_21-04-27
294020-hifi-snap-by--on-2017-11-08_21-04-28
294020-hifi-snap-by--on-2017-11-08_21-04-29
294020-hifi-snap-by--on-2017-11-08_21-04-30
294020-hifi-snap-by--on-2017-11-08_21-04-31
294020-hifi-snap-by--on-2017-11-08_21-04-32
294020-hifi-snap-by--on-2017-11-08_21-04-33
294020-hifi-snap-by--on-2017-11-08_21-04-34
294020-hifi-snap-by--on-2017-11-08_21-04-35
294020-hifi-snap-by--on-2017-11-08_21-04-36
294020-hifi-snap-by--on-2017-11-08_21-04-37
294020-hifi-snap-by--on-2017-11-08_21-04-38
294020-hifi-snap-by--on-2017-11-08_21-04-39
294020-hifi-snap-by--on-2017-11-08_21-04-40
294020-hifi-snap-by--on-2017-11-08_21-04-41
294020-hifi-snap-by--on-2017-11-08_21-04-42

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4803/HighFidelity-Beta-PR11650-4f7ed38e98e2c49a3764a608dabaee7cbb8bfaaa.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4803/HighFidelity-Beta-PR11650-4f7ed38e98e2c49a3764a608dabaee7cbb8bfaaa.dmg

**Comment by Nex-Pro** [comment]
Retesting with most recent build.

**Comment by Nex-Pro** [comment]
### PASS All desktop and HMD outlines are as required.

@Zvork Good job. 

PS. We probably want to delete any unneeded images from this page. ( probably all of them ) 

io3472-hifi-snap-by--on-2017-11-10_18-25-03
aj3472-hifi-snap-by--on-2017-11-10_18-24-59
fe3472-hifi-snap-by--on-2017-11-10_18-24-53
yg3472-hifi-snap-by--on-2017-11-10_18-24-49
eq3472-hifi-snap-by--on-2017-11-10_18-24-45
td3472-hifi-snap-by--on-2017-11-10_18-24-42
nt3472-hifi-snap-by--on-2017-11-10_18-24-38
tp3472-hifi-snap-by--on-2017-11-10_18-24-32
tb3472-hifi-snap-by--on-2017-11-10_18-24-29
vk3472-hifi-snap-by--on-2017-11-10_18-24-25
cr3472-hifi-snap-by--on-2017-11-10_18-24-22
yw3472-hifi-snap-by--on-2017-11-10_18-24-17
sg3472-hifi-snap-by--on-2017-11-10_18-24-11
on3472-hifi-snap-by--on-2017-11-10_18-24-06
rn3472-hifi-snap-by--on-2017-11-10_18-23-57
ir3472-hifi-snap-by--on-2017-11-10_18-23-44
ya3472-hifi-snap-by--on-2017-11-10_18-23-34

### HMD

hifi-snap-by-piper-peppercorn-on-2017-11-10_18-36-36
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-36-42
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-36-45
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-36-48
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-36-51
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-36-53
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-36-55
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-36-57
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-37-00
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-37-02
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-37-04
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-37-06
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-37-08
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-37-10
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-37-12
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-37-15
hifi-snap-by-piper-peppercorn-on-2017-11-10_18-37-17
